File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = [(v, k) for k,v in Counter(nums).items()]
        hp=[]
        for i in freq:
            heappush(hp, i)
            if len(hp)>k:
                heappop(hp)
        return [k for v,k in hp]
        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        var map = new HashMap<Integer, Integer>();
        for(var n: nums)
            map.put(n, map.getOrDefault(n, 0) + 1);
        
        var pq = new PriorityQueue<int[]>(Comparator.comparingInt(a -> a[0]));
        for(var key : map.keySet()){
            pq.offer(new int[]{map.get(key), key});
            if(pq.size() > k)
                pq.poll();
        }
        
        int[] ans = new int[k];
        int ptr = 0;
        
        for(var x: pq)
            ans[ptr++] = x[1];
        
        return ans;
    }
}